msal/src/main/java/com/microsoft/identity/client/PublicClientApplication.java

```diff
@@ -554,51 +557,127 @@ public void run() {
     }
 
     /**
-     * Returns the IAccount object matching the supplied home_account_id.
+     * Returns the IAccount object matching the supplied object id.
      *
-     * @param homeAccountIdentifier The home_account_id of the sought IAccount.
-     * @param authority             The authority of the sought IAccount.
-     * @return The IAccount stored in the cache or null, if no such matching entry exists.
+     * @param oid The object id of the sought IAccount.
+     * @param callback AccountLoadedCallback
      */
-    @Nullable
-    public IAccount getAccount(@NonNull final String homeAccountIdentifier,
-                               @Nullable final String authority) {
+    public void getAccount(@NonNull final String oid,
```

Comment:
I think that the factoring of this method is more complicated than it needs to be....

```java
void getAccount(@NonNull final String oid, @NonNull final String AccountsLoadedCallback<IAccount> callback) {
    final AccountsLoadedCallback<IAccount> brokerCallback = new AccountsLoadedCallback<>() {
        @Override
        public void onAccountsLoaded(@Nullable final IAccount account) {
            callback.onAccountsLoaded(account);
        }
    }

    getLocalAccountByOid(oid, new AccountsLoadedCallback<IAccount>() {
        @Override
        public void onAccountsLoaded(@Nullable final IAccount account) {
            if (null == account) {
                getBrokerAccountByOid(oid, brokerCallback); // Method yields null if not broker elligible
            } else {
                callback.onAccountsLoaded(account);
            }
        }
    })
}
```

Then.... we define two methods...

```java
// One method to load accounts locally and return a List
private void getLocalAccountByOid(@NonNull String oid, @NonNull final AccountsLoadedCallback<IAccount> callback) {
    // On a worker thread, if necessary....
    final List<IAccount> accounts = getLocalAccounts();
    callback.onAccountsLoaded(getAccountByOid(acounts));
}
````

````java
// Another method to load Accounts from Broker and return a List
private void getBrokerAccountByOid(@NonNull String oid, @NonNull final AccountsLoadedCallback<IAccount> callback) {
    // On a worker thread, if necessary....
    final List<IAccount> accounts = getBrokerAccounts(); // null or empty, as appropriate (null for no-broker, empty for no-accounts)
    callback.onAccountsLoaded(getAccountByOid(accounts));
}
````

And getAccountByOid() does all the filtering once....

```java
public IAccount getAccountByOid(@NonNull final String oid, @NonNull List<IAccount> accounts) {
    // fill in the rest here....
}
````

Reply:
I'm afraid the getBrokerAccountByOid cannot be refactored in this way. Because the call to broker cannot happen on main thread like the getLocalAccount, so it has to call the looper to post the result on a different thread. Thus, it is better to get both broker and local accounts on the same thread, fetch the target account and return it back.